File: import_vapi_calls.mjs
#!/usr/bin/env node
/**
 * import_vapi_calls.mjs
 * 
 * Fetches all calls from Vapi API and imports them into NocoDB call_logs table.
 * Test/manual calls (without empresa) are tagged as 'Manual Trigger'.
 */

const VAPI_API_KEY = '0594f41c-e836-425d-aaa2-1c5b7d9e506e';
const NOCODB_BASE = 'https://optima-nocodb.vhsxer.easypanel.host/api/v2/tables';
const CALL_LOGS_TABLE = 'm73w58ba47ifkrx';
const XC_TOKEN = 'vodwktZQ77mth3XeK290Fw8V9Axloe1LiOxsWn5d';

async function fetchVapiCalls() {
    const res = await fetch('https://api.vapi.ai/call?limit=100', {
        headers: { 'Authorization': `Bearer ${VAPI_API_KEY}` }
    });
    if (!res.ok) throw new Error(`Vapi API error: ${res.status} ${await res.text()}`);
    return res.json();
}

function calcDuration(call) {
    if (call.startedAt && call.endedAt) {
        return Math.round((new Date(call.endedAt) - new Date(call.startedAt)) / 1000);
    }
    return 0;
}

function evaluateCall(ended, duration) {
    if (ended === 'voicemail') return 'Contestador';
    if (ended === 'silence-timed-out') return 'No contesta';
    if (ended === 'customer-busy') return 'No contesta';
    if (ended && ended.includes('error')) return 'Error';
    if (duration && duration < 10) return 'No contesta';
    if (ended === 'customer-ended-call' && duration > 30) return 'Completada';
    if (ended === 'assistant-ended-call' && duration > 20) return 'Completada';
    return 'Sin datos';
}

async function importToNocoDB(records) {
    const batchSize = 10;
    let success = 0;

    for (let i = 0; i < records.length; i += batchSize) {
        const batch = records.slice(i, i + batchSize);
        try {
            const res = await fetch(`${NOCODB_BASE}/${CALL_LOGS_TABLE}/records`, {
                method: 'POST',
                headers: {
                    'xc-token': XC_TOKEN,
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(batch)
            });
            if (!res.ok) {
                const errBody = await res.text();
                console.log(`  Batch ${Math.floor(i / batchSize) + 1}: ERROR ${res.status} - ${errBody.substring(0, 300)}`);
                continue;
            }
            const result = await res.json();
            const count = Array.isArray(result) ? result.length : 1;
            success += count;
            console.log(`  Batch ${Math.floor(i / batchSize) + 1}: imported ${count} records`);
        } catch (err) {
            console.log(`  Batch ${Math.floor(i / batchSize) + 1}: ERROR - ${err.message}`);
        }
    }
    return success;
}

async function main() {
    console.log('📡 Fetching calls from Vapi API...');
    const calls = await fetchVapiCalls();
    console.log(`   Found ${calls.length} calls in Vapi\n`);

    const records = [];
    let testCount = 0;
    let campaignCount = 0;

    for (const c of calls) {
        const empresa = c.assistantOverrides?.variableValues?.empresa || '';
        const phone = c.customer?.number || 'Desconocido';
        const created = c.createdAt || '';
        const ended = c.endedReason || '';
        const callId = c.id || '';
        const transcript = c.artifact?.transcript || '';
        const recording = c.artifact?.recordingUrl || '';
        const isTest = !empresa;
        const duration = calcDuration(c);
        const evaluation = evaluateCall(ended, duration);

        const record = {
            vapi_call_id: callId,
            lead_name: isTest ? 'Test Manual' : empresa,
            phone_called: phone,
            call_time: created,
            ended_reason: isTest ? 'Manual Trigger' : ended,
            evaluation: evaluation,
            duration_seconds: duration,
        };

        if (transcript) record.transcript = transcript.substring(0, 5000);
        if (recording) record.recording_url = recording;

        records.push(record);
        if (isTest) testCount++;
        else campaignCount++;
    }

    console.log('📊 Records prepared:');
    console.log(`   🧪 Test (manual): ${testCount}`);
    console.log(`   📞 Campaign: ${campaignCount}`);
    console.log(`   📦 Total: ${records.length}\n`);

    console.log('⬆️  Importing to NocoDB...');
    const success = await importToNocoDB(records);
    console.log(`\n✅ Import complete: ${success}/${records.length} records imported`);
}

main().catch(err => {
    console.error('❌ Error:', err.message);
    process.exit(1);
});

File: import_vapi_calls.py
#!/usr/bin/env python3
"""
Import all Vapi calls into NocoDB call_logs table.
Test calls (no empresa) are marked with ended_reason='Manual Trigger'.
Campaign calls keep their original ended_reason.
"""

import json
import urllib.request
from datetime import datetime

VAPI_API_KEY = "852080ba-ce7c-4778-b218-bf718613a2b6"
NOCODB_BASE = "https://nocodb.srv889387.hstgr.cloud/api/v2/tables"
CALL_LOGS_TABLE = "m013en5u2cyu30j"
XC_TOKEN = "jx3uoKeVaidZLF7M0skVb9pV6yvNsam0Hu-Vfeww"

def fetch_vapi_calls():
    req = urllib.request.Request(
        "https://api.vapi.ai/call?limit=100",
        headers={"Authorization": f"Bearer {VAPI_API_KEY}"}
    )
    resp = urllib.request.urlopen(req)
    return json.loads(resp.read())

def calc_duration(call):
    if call.get('startedAt') and call.get('endedAt'):
        s = datetime.fromisoformat(call['startedAt'].replace('Z', '+00:00'))
        e = datetime.fromisoformat(call['endedAt'].replace('Z', '+00:00'))
        return int((e - s).total_seconds())
    return 0

def evaluate_call(ended, duration):
    if ended == 'voicemail':
        return 'Contestador'
    if ended == 'silence-timed-out':
        return 'No contesta'
    if ended == 'customer-busy':
        return 'No contesta'
    if ended and 'error' in ended:
        return 'Error'
    if duration and duration < 10:
        return 'No contesta'
    if ended == 'customer-ended-call' and duration and duration > 30:
        return 'Completada'
    if ended == 'assistant-ended-call' and duration and duration > 20:
        return 'Completada'
    return 'Sin datos'

def import_to_nocodb(records):
    batch_size = 10
    success = 0
    for i in range(0, len(records), batch_size):
        batch = records[i:i+batch_size]
        url = f"{NOCODB_BASE}/{CALL_LOGS_TABLE}/records"
        payload = json.dumps(batch).encode('utf-8')
        
        req = urllib.request.Request(url, data=payload, method='POST')
        req.add_header('xc-token', XC_TOKEN)
        req.add_header('Content-Type', 'application/json')
        
        try:
            resp = urllib.request.urlopen(req)
            result = json.loads(resp.read())
            created_count = len(result) if isinstance(result, list) else 1
            success += created_count
            print(f"  Batch {i//batch_size + 1}: imported {created_count} records")
        except Exception as e:
            error_body = ''
            if hasattr(e, 'read'):
                error_body = e.read().decode('utf-8', errors='replace')
            print(f"  Batch {i//batch_size + 1}: ERROR - {e}")
            print(f"    Body: {error_body[:500]}")
    
    return success

def main():
    print("📡 Fetching calls from Vapi API...")
    calls = fetch_vapi_calls()
    print(f"   Found {len(calls)} calls in Vapi\n")
    
    records = []
    test_count = 0
    campaign_count = 0
    
    for c in calls:
        empresa = c.get('assistantOverrides', {}).get('variableValues', {}).get('empresa', '')
        phone = c.get('customer', {}).get('number', '') or 'Desconocido'
        created = c.get('createdAt', '')
        ended = c.get('endedReason', '')
        call_id = c.get('id', '')
        transcript = c.get('artifact', {}).get('transcript', '') or ''
        recording = c.get('artifact', {}).get('recordingUrl', '') or ''
        
        is_test = not empresa
        duration = calc_duration(c)
        evaluation = evaluate_call(ended, duration)
        
        record = {
            'vapi_call_id': call_id,
            'lead_name': 'Test Manual' if is_test else empresa,
            'phone_called': phone,
            'call_time': created,
            'ended_reason': 'Manual Trigger' if is_test else ended,
            'evaluation': evaluation,
            'duration_seconds': duration,
        }
        if transcript:
            record['transcript'] = transcript[:5000]
        if recording:
            record['recording_url'] = recording
        
        records.append(record)
        if is_test:
            test_count += 1
        else:
            campaign_count += 1
    
    print(f"📊 Records prepared:")
    print(f"   🧪 Test (manual): {test_count}")
    print(f"   📞 Campaign: {campaign_count}")
    print(f"   📦 Total: {len(records)}\n")
    
    print("⬆️  Importing to NocoDB...")
    success = import_to_nocodb(records)
    print(f"\n✅ Import complete: {success}/{len(records)} records imported")

if __name__ == '__main__':
    main()
